synapse/config/saml2_config.py

```diff
@@ -12,11 +13,29 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
 from synapse.python_dependencies import DependencyException, check_requirements
+from synapse.util.module_loader import load_python_module
 
 from ._base import Config, ConfigError
 
 
+def _dict_merge(merge_dict, into_dct):
```

Comment:
Docstrings please. This looks to be attempting a deep merge? Is that what we want?

Reply:
yes, and yes I think so. The config dict is quite deeply nested and needs sensible defaults at various levels.

Reply:
This is probably fine in practice, though does mean that e.g. if you merged two configs the specified different `organization` dicts then that could lead to interesting results if they didn't set the same fields.

Reply:
true, though again: this is more about merging the defaults and the custom settings, rather than merging an inline config with a separate file. If you're using the inline config and a separate file, I think you get to keep both pieces when it breaks.